test/e2e/autoscale_hpa_test.go

```diff
@@ -0,0 +1,205 @@
+				autoscaling.MetricAnnotationKey:   metric,
+				autoscaling.TargetAnnotationKey:   strconv.Itoa(target),
+				autoscaling.MinScaleAnnotationKey: fmt.Sprintf("%d", minimumReplicas),
+				autoscaling.MaxScaleAnnotationKey: fmt.Sprintf("%d", int(maxPods)),
+			}), rtesting.WithResourceRequirements(corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("30m"),
+					corev1.ResourceMemory: resource.MustParse("20Mi"),
+				},
+			}),
+		}...)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	if _, err := pkgTest.WaitForEndpointState(
+		context.Background(),
+		clients.KubeClient,
+		t.Logf,
+		names.URL,
+		v1test.RetryingRouteInconsistency(spoof.MatchesAllOf(spoof.IsStatusOK)),
+		"CheckingEndpointAfterCreate",
+		false,
+		test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS),
+	); err != nil {
+		t.Fatalf("Error probing %s: %v", names.URL.Hostname(), err)
+	}
+
+	return &TestContext{
+		t:           t,
+		logf:        t.Logf,
+		clients:     clients,
+		names:       names,
+		resources:   resources,
+		targetValue: target,
+		metric:      metric,
+	}
+}
+func assertHPAAutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) {
+	ctx.t.Helper()
+	wait := autoscaleHPAUpToNumPods(ctx, curPods, targetPods, done, quick)
+	if err := wait(); err != nil {
+		ctx.t.Fatal(err)
+	}
+}
+
+func autoscaleHPAUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) func() error {
+	ctx.t.Helper()
+
+	stopChan := make(chan struct{})
+	var grp errgroup.Group
+	grp.Go(func() error {
+		return generateTrafficAtFixedRPSWithCPULoad(ctx, int(targetPods*5), stopChan)
+	})
+
+	grp.Go(func() error {
+		defer close(stopChan)
+		return checkPodScale(ctx, targetPods, minPods, maxPods, done, quick)
+	})
+
+	return grp.Wait
+}
+
+func generateTrafficAtFixedRPSWithCPULoad(ctx *TestContext, rps int, stopChan chan struct{}) error {
+	pacer := vegeta.ConstantPacer{Freq: rps, Per: time.Second}
+	attacker := vegeta.NewAttacker(vegeta.Timeout(0)) // No timeout is enforced at all.
+	target, err := getVegetaTarget(
+		ctx.clients.KubeClient, ctx.resources.Route.Status.URL.URL().Hostname(), pkgTest.Flags.IngressEndpoint, test.ServingFlags.ResolvableDomain, "prime", primeNum)
+	if err != nil {
+		return fmt.Errorf("error creating vegeta target: %w", err)
+	}
+	ctx.logf("Maintaining %v RPS.", rps)
+	return generateTraffic(ctx, attacker, pacer, stopChan, target)
+}
+
+func assertScaleDownToOne(ctx *TestContext) {
+	deploymentName := resourcenames.Deployment(ctx.resources.Revision)
+	if err := waitForScaleToOne(ctx.t, deploymentName, ctx.clients); err != nil {
+		ctx.t.Fatalf("Unable to observe the Deployment named %s scaling down: %v", deploymentName, err)
+	}
+	// Account for the case where scaling up uses all available pods.
```

Comment:
What's this comment supposed to mean?

Reply:
I guess it is a relic from the kpa logic, will remove.